Optimizing TiDB Settings for Maximum Performance (#19029)
pingcap/docs: performance-best-practice.md

# Why This Matters

Optimizing TiDB for peak performance requires careful tuning of various settings. In many cases, achieving optimal performance involves adjusting configurations beyond their default values.

While default settings are chosen for stability, maximizing performance often requires more aggressive configurations and sometimes the use of experimental features. This approach is based on insights gained from numerous real-world deployments and performance optimization efforts.

This guide explains these non-default settings, detailing their benefits and potential trade-offs. It's designed to help you make informed decisions when fine-tuning TiDB for your specific workload requirements.

# Key Settings for Common Workloads

The following suggested settings cover the most common optimizations for improving TiDB performance:

- Enhancing query plan caching
- Optimizing the query optimizer's behavior
- More aggressive use of TiKV's Titan storage engine

These settings can significantly boost performance for many common workloads, but as with any optimization, it's important to test thoroughly in your specific environment.

## system variables

> ```SQL 

Reviewer: The `tidb_enable_stmt_summary` impacts the OLTP performance a lot and it would consume a lot of cpu resouces, please ensure it's turned off in exterm OLTP benchmark performance test.

Author: disable tidb_enable_stmt_summary is too extreme for general PoC.

- 44830:ON: Plan Cache is allowed to cache execution plans with the PointGet operator generated during physical optimization.
- 44855:ON: Enable IndexJoin when the Probe side of an IndexJoin operator contains a Selection operator.
- 52869:ON: Enable indexmerge if the optimizer can choose the single index scan method (other than full table scan) for a query plan

## TiDB Configurations

```toml
[performance]
concurrently-init-stats: true
force-init-stats: true
lite-init-stats: false

[tikv-client]
region-cache-ttl = 1200
```

### Justifications

| configurations | Pro | Cons | 
| ---------| ---- | ----|
| concurrently-init-stats force-init-stats lite-init-stats| make sure the statistics is full loaded and ready during TiDB startup | need to wait more time during TiDB startup and more memory usage | 
| region-cache-ttl| Increase from 10m to 20m, to reduce the GetRegion request to pd server | Higher possibility for leaders/region miss backoff if the data are changed and moved frequently | 

## TiKV Configurations

```toml
[server]

> concurrent-send-snap-limit = 64

Author: it's hard to adjust concurrent-send-snap-limit according to the instance type, so we just double the default value

---
title: Optimizing TiDB Settings for Maximum Performance
summary: Learn how to optimize TiDB performance by configuring key settings and addressing edge cases
---

# Optimizing TiDB Settings for Maximum Performance

This guide provides essential information on optimizing TiDB for maximum performance. We'll explore:
- Performance best practices for common workloads
- Strategies for addressing challenging performance scenarios

> > **Important Note:**

Reviewer 2: ```suggestion
> **Note:**
```

# Why This Matters

Optimizing TiDB for peak performance requires careful tuning of various settings. In many cases, achieving optimal performance involves adjusting configurations beyond their default values.

While default settings are chosen for stability, maximizing performance often requires more aggressive configurations and sometimes the use of experimental features. This approach is based on insights gained from numerous real-world deployments and performance optimization efforts.

This guide explains these non-default settings, detailing their benefits and potential trade-offs. It's designed to help you make informed decisions when fine-tuning TiDB for your specific workload requirements.

# Key Settings for Common Workloads

The following suggested settings cover the most common optimizations for improving TiDB performance:

- Enhancing query plan caching
- Optimizing the query optimizer's behavior
- More aggressive use of TiKV's Titan storage engine

These settings can significantly boost performance for many common workloads, but as with any optimization, it's important to test thoroughly in your specific environment.

> ## system variables

Reviewer 2: ```suggestion
## System variables
```
Please use sentence case (capitalize only the first word and proper nouns) for all headings at H2 level and below.

#### fix control explanation
Use aggressive optimizer controls to enable more possible optimizations.

- 44262:ON: Allow the use of Dynamic pruning mode to access the partitioned table when the GlobalStats are missing.
- 44389:ON: For filters such as c = 10 and (a = 'xx' or (a = 'kk' and b = 1)), this variable enable to try to build more comprehensive scan ranges for IndexRangeScan.
- 44823:10000: To save memory, Plan Cache does not cache queries with parameters exceeding the specified number of this variable. Increased from from 200 to 10000 to make plan cache availabe for query with long in-list;
- 44830:ON: Plan Cache is allowed to cache execution plans with the PointGet operator generated during physical optimization.
- 44855:ON: Enable IndexJoin when the Probe side of an IndexJoin operator contains a Selection operator.
- 52869:ON: Enable indexmerge if the optimizer can choose the single index scan method (other than full table scan) for a query plan

## TiDB Configurations

```toml
[performance]

> concurrently-init-stats: true
> force-init-stats: true
> lite-init-stats: false

Reviewer 2: ```suggestion
concurrently-init-stats = true
force-init-stats = true
lite-init-stats = false
```

| rocksdb.titan rocksdb.defaultcf.titan | RocksDB might exhibit high write amplification, and the disk throughput might become the bottleneck for the workload. As a result, the total number of pending compaction bytes grows over time and triggers flow control, which indicates that TiKV lacks sufficient disk bandwidth to keep up with the foreground write flow.  To alleviate the bottleneck caused by limited disk throughput, you can improve performance by enabling Titan. | When Titan is enabled, there might be a slight performance degradation for range scans on the primary key, and the space amplification is higher than RocksDB, at worse case, the space usage is 2x of the original data. For more information, see Impact of min-blob-size on performance. | 
| l0-files-threshold | increase the thrshhold to avoid unneccessary flow control | - | 

> ## TiFlash Congiurations

Reviewer 2: ```suggestion
## TiFlash configurations
```

> ---
> title: Optimizing TiDB Settings for Maximum Performance
> summary: Learn how to optimize TiDB performance by configuring key settings and addressing edge cases
> ---

Reviewer 2: ```suggestion
---
title: Configure TiDB for Optimal Performance
summary: Learn how to optimize TiDB performance by configuring key settings and addressing edge cases.
---
```

> ## Optimizing TiDB Settings for Maximum Performance

Reviewer 2: ```suggestion
## Configure TiDB for Optimal Performance
```

## Why This Matters

Optimizing TiDB for peak performance requires careful tuning of various settings. In many cases, achieving optimal performance involves adjusting configurations beyond their default values.

While default settings are chosen for stability, maximizing performance often requires more aggressive configurations and sometimes the use of experimental features. This approach is based on insights gained from numerous real-world deployments and performance optimization efforts.

This guide explains these non-default settings, detailing their benefits and potential trade-offs. It's designed to help you make informed decisions when fine-tuning TiDB for your specific workload requirements.

## Key Settings for Common Workloads

The following suggested settings cover the most common optimizations for improving TiDB performance:

> - Enhancing query plan caching
> - Optimizing the query optimizer's behavior
> - More aggressive use of TiKV's Titan storage engine

Reviewer 2: ```suggestion
- Enhance execution plan caching, such as [Instance-level execution plan cache](/sql-prepared-plan-cache.md), [Non-prepared plan cache](/sql-non-prepared-plan-cache.md), and [Instance-level execution plan cache](/system-variables.md#tidb_enable_instance_plan_cache-new-in-v840).
- Optimize behavior of the TiDB optimizer using [Optimizer Fix Controls](/optimizer-fix-controls.md).
- Use the [Titan](/storage-engine/titan-overview.md) storage engine more aggressively.
```

> These settings can significantly boost performance for many common workloads, but as with any optimization, it's essential to test thoroughly in your specific environment.

Reviewer 2: ```suggestion
These settings can significantly improve performance for many workloads. However, as with any optimization, thoroughly test them in your environment before deploying to production.
```

### System variables

> 

Reviewer 2: ```suggestion

Execute the following SQL commands to apply the recommended settings:

```

> ```SQL 
> set global tidb_enable_instance_plan_cache=on;
> set global tidb_instance_plan_cache_max_size=2GiB;
> set global tidb_enable_non_prepared_plan_cache=on;
> set global tidb_ignore_prepared_cache_close_stmt=on;
> set global tidb_enable_inl_join_inner_multi_pattern=on;
> set global tidb_opt_derive_topn=on;
> set global tidb_runtime_filter_mode=LOCAL;
> set global tidb_opt_enable_mpp_shared_cte_execution=on;
> set global tidb_rc_read_check_ts=on;
> set global tidb_guarantee_linearizability=off;
> set global pd_enable_follower_handle_region=on;
> set global tidb_opt_fix_control = '44262:ON,44389:ON,44823:10000,44830:ON,44855:ON,52869:ON';
> ```

Reviewer 2: ````suggestion
```sql
SET GLOBAL tidb_enable_instance_plan_cache=on;
SET GLOBAL tidb_instance_plan_cache_max_size=2GiB;
SET GLOBAL tidb_enable_non_prepared_plan_cache=on;
SET GLOBAL tidb_ignore_prepared_cache_close_stmt=on;
SET GLOBAL tidb_enable_inl_join_inner_multi_pattern=on;
SET GLOBAL tidb_opt_derive_topn=on;
SET GLOBAL tidb_runtime_filter_mode=LOCAL;
SET GLOBAL tidb_opt_enable_mpp_shared_cte_execution=on;
SET GLOBAL tidb_rc_read_check_ts=on;
SET GLOBAL tidb_guarantee_linearizability=off;
SET GLOBAL pd_enable_follower_handle_region=on;
SET GLOBAL tidb_opt_fix_control = '44262:ON,44389:ON,44823:10000,44830:ON,44855:ON,52869:ON';
```
````

> #### Justifications

Reviewer 2: ```suggestion
The following table outlines the impact of specific system variable configurations:
```

> | variables| Pro | Cons | 

Reviewer 2: ```suggestion
| System variable | Description | Note |
```
In technical documentation, it’s generally better to avoid informal or overly casual terms like "pro" and "cons".

| ---------| ---- | ----|

> | [`tidb_enable_instance_plan_cache`](/system-variables.md#tidb_enable_instance_plan_cache-new-in-v840) [`tidb_instance_plan_cache_max_size`](/system-variables.md#tidb_instance_plan_cache_max_size-new-in-v840)| Introduces instance-level plan caching to replace session-level caching, significantly improving performance for workloads with high connection counts or frequent prepared statement usage | As an experimental feature, carefully evaluate in non-production environments first. Monitor memory usage as plan cache size increases |

Reviewer 2: ```suggestion
| [`tidb_enable_instance_plan_cache`](/system-variables.md#tidb_enable_instance_plan_cache-new-in-v840) and [`tidb_instance_plan_cache_max_size`](/system-variables.md#tidb_instance_plan_cache_max_size-new-in-v840)| Use instance-level plan cache instead of session-level caching. This significantly improves performance for workloads with high connection counts or frequent prepared statement usage. | This is an experimental feature. Test in non-production environments first and monitor memory usage as the plan cache size increases. |
```

> | [`tidb_enable_non_prepared_plan_cache`](/system-variables.md#tidb_enable_non_prepared_plan_cache)| Turns on none prepared plan cache for applications not using prepared statement to reduce the compile cost | - | 

Reviewer 2: ```suggestion
| [`tidb_enable_non_prepared_plan_cache`](/system-variables.md#tidb_enable_non_prepared_plan_cache)| Enable the [Non-prepared plan cache](/sql-non-prepared-plan-cache.md) feature to reduce compile costs for applications that do not use prepared statements. | N/A |
```

> | [`tidb_ignore_prepared_cache_close_stmt`](/system-variables.md#tidb_ignore_prepared_cache_close_stmt-new-in-v600)| Caches plan for applications using prepared statements but close the plan after every single execution | - | 

Reviewer 2: ```suggestion
| [`tidb_ignore_prepared_cache_close_stmt`](/system-variables.md#tidb_ignore_prepared_cache_close_stmt-new-in-v600)| Cache plans for applications that use prepared statements but close the plan after each execution. | N/A |
```

> | [`tidb_enable_inl_join_inner_multi_pattern`](/system-variables.md#tidb_enable_inl_join_inner_multi_pattern-new-in-v700)| Index Join is supported when the inner table has Selection or Projection operators on it | - | 

Reviewer 2: ```suggestion
| [`tidb_enable_inl_join_inner_multi_pattern`](/system-variables.md#tidb_enable_inl_join_inner_multi_pattern-new-in-v700)| Enable Index Join support when the inner table has `Selection` or `Projection` operators on it. | N/A | 
```

> | [`tidb_opt_derive_topn`](/system-variables.md#tidb_opt_derive_topn-new-in-v700)| Enables the optimization rule of Deriving TopN or Limit from window functions | Only limited to ROW_NUMBER() window function | 

Reviewer 2: ```suggestion
| [`tidb_opt_derive_topn`](/system-variables.md#tidb_opt_derive_topn-new-in-v700)| Enable the optimization rule of [Deriving TopN or Limit from window functions](/derive-topn-from-window.md). | This is limited to the `ROW_NUMBER()` window function. | 
```

> | [`tidb_runtime_filter_mode`](/system-variables.md#tidb_runtime_filter_mode-new-in-v720)| Enables runtime filter to improve the hash join efficiency | The variable is added to v7.2.0, and for safety, it's disabled by default | 

Reviewer 2: ```suggestion
| [`tidb_runtime_filter_mode`](/system-variables.md#tidb_runtime_filter_mode-new-in-v720)| Enable [Runtime Filter](/runtime-filter.md#runtime-filter-mode) in the local mode to improve hash join efficiency. | The variable is introduced in v7.2.0 and is disabled by default for safety. | 
```

> | [`tidb_opt_enable_mpp_shared_cte_execution`](/system-variables.md#tidb_opt_enable_mpp_shared_cte_execution-new-in-v720)| Enables non-recursive Common Table Expressions (CTE) pushdown to TiFlash | The feature is experimental  | 

Reviewer 2: ```suggestion
| [`tidb_opt_enable_mpp_shared_cte_execution`](/system-variables.md#tidb_opt_enable_mpp_shared_cte_execution-new-in-v720)| Enable non-recursive [Common Table Expressions (CTE)](/sql-statements/sql-statement-with.md) pushdown to TiFlash. | This is an experimental feature. | 
```

> | [`tidb_rc_read_check_ts`](/system-variables.md#tidb_rc_read_check_ts-new-in-v600)| For read-committed isolation level, Enables this variable can avoid the latency and cost of getting the global timestamp, and can optimize the transaction-level read latency. | This feature is incompatible with repeatable-read | 

Reviewer 2: ```suggestion
| [`tidb_rc_read_check_ts`](/system-variables.md#tidb_rc_read_check_ts-new-in-v600)| For the read-committed isolation level, enabling this variable avoids the latency and cost of getting the global timestamp and optimizes transaction-level read latency. | This feature is incompatible with the Repeatable Read isolation level. | 
```

> | [`tidb_guarantee_linearizability`](/system-variables.md#tidb_guarantee_linearizability-new-in-v50)| Improves performance by skipping the commit timestamp fetch from PD server | Trades linearizability for performance - only causal consistency is guaranteed. Not suitable for scenarios requiring strict linearizability |

Reviewer 2: ```suggestion
| [`tidb_guarantee_linearizability`](/system-variables.md#tidb_guarantee_linearizability-new-in-v50)| Improve performance by skipping the commit timestamp fetch from the PD server. | This sacrifices linearizability in favor of performance. Only causal consistency is guaranteed. It is not suitable for scenarios requiring strict linearizability. |
```

> | [`pd_enable_follower_handle_region`](/system-variables.md#pd_enable_follower_handle_region-new-in-v760)| Activates the PD Follower feature, allowing PD followers to process Region requests, which helps distribute load evenly across all PD servers and reduces CPU pressure on the PD leader | This feature is experimental and should be tested in non-production environments |

Reviewer 2: ```suggestion
| [`pd_enable_follower_handle_region`](/system-variables.md#pd_enable_follower_handle_region-new-in-v760)| Activate the PD Follower feature, allowing PD followers to process Region requests. This helps distribute load evenly across all PD servers and reduces CPU pressure on the PD leader. | This is an experimental feature. Test in non-production environments. |
```

> | [`tidb_opt_fix_control`](/optimizer-fix-controls.md#tidb_opt_fix_control) | Enables advanced query optimization strategies to improve performance through additional optimization rules and heuristics | Test thoroughly in your environment as benefits vary by workload |

Reviewer 2: ```suggestion
| [`tidb_opt_fix_control`](/optimizer-fix-controls.md#tidb_opt_fix_control) | Enable advanced query optimization strategies to improve performance through additional optimization rules and heuristics. | Test thoroughly in your environment, as performance improvements vary by workload. |
```

> #### fix control explanation
> 
> Use aggressive optimizer controls to enable more possible optimizations.

Reviewer 2: ```suggestion
The following describes the optimizer control configurations that enable additional optimizations:
```

> - [`44262`](/optimizer-fix-controls.md#44262-new-in-v653-and-v720):ON: Allow the use of Dynamic pruning mode to access the partitioned table when the GlobalStats are missing.
> - [`44389`](/optimizer-fix-controls.md#44389-new-in-v653-and-v720):ON: For filters such as c = 10 and (a = 'xx' or (a = 'kk' and b = 1)), this variable enable to try to build more comprehensive scan ranges for IndexRangeScan.
> - [`44823`](/optimizer-fix-controls.md#44823-new-in-v730):10000: To save memory, Plan Cache does not cache queries with parameters exceeding the specified number of this variable. Increased from from 200 to 10000 to make plan cache availabe for query with long in-list;
> - [`44830`](/optimizer-fix-controls.md#44830-new-in-v730):ON: Plan Cache is allowed to cache execution plans with the PointGet operator generated during physical optimization.
> - [`44855`](/optimizer-fix-controls.md#44855-new-in-v730):ON: Enable IndexJoin when the Probe side of an IndexJoin operator contains a Selection operator.
> - [`52869`](/optimizer-fix-controls.md#52869-new-in-v810):ON: Enable indexmerge if the optimizer can choose the single index scan method (other than full table scan) for a query plan

Reviewer 2: ```suggestion
- [`44262:ON`](/optimizer-fix-controls.md#44262-new-in-v653-and-v720): Use [Dynamic pruning mode](/partitioned-table.md#dynamic-pruning-mode) to access the partitioned table when the [GlobalStats](/statistics.md#collect-statistics-of-partitioned-tables-in-dynamic-pruning-mode) are missing.
- [`44389:ON`](/optimizer-fix-controls.md#44389-new-in-v653-and-v720): For filters such as `c = 10 and (a = 'xx' or (a = 'kk' and b = 1))`, build more comprehensive scan ranges for `IndexRangeScan`.
- [`44823:10000`](/optimizer-fix-controls.md#44823-new-in-v730): To save memory, Plan Cache does not cache queries with parameters exceeding the specified number of this variable. Increase plan cache parameter limit from `200` to `10000` to make plan cache available for query with long in-lists.
- [`44830:ON`](/optimizer-fix-controls.md#44830-new-in-v730): Plan Cache is allowed to cache execution plans with the `PointGet` operator generated during physical optimization.
- [`44855:ON`](/optimizer-fix-controls.md#44855-new-in-v730): The optimizer selects `IndexJoin` when the `Probe` side of an `IndexJoin` operator contains a `Selection` operator.
- [`52869:ON`](/optimizer-fix-controls.md#52869-new-in-v810): The optimizer chooses index merge automatically if the optimizer can choose the single index scan method (other than full table scan) for a query plan.
```

> ### TiDB Configurations

Reviewer 2: ```suggestion
### TiDB configurations
```

> 

Reviewer 2: ```suggestion

Add the following configuration items to the TiDB configuration file:

```

```toml
[performance]
concurrently-init-stats = true
force-init-stats = true
lite-init-stats = false
```

> #### Justifications

Reviewer 2: ```suggestion
```
Remove this heading, as it is the only L4 heading in this section

> | configurations | Pro | Cons | 
> | ---------| ---- | ----|
> | [`concurrently-init-stats`](/tidb-configuration-file.md#concurrently-init-stats-new-in-v810-and-v752) [`force-init-stats`](/tidb-configuration-file.md#force-init-stats-new-in-v657-and-v710) [`lite-init-stats`](/tidb-configuration-file.md#lite-init-stats-new-in-v710) | Ensures comprehensive and concurrent loading of table statistics during TiDB startup, improving query optimization readiness | Potentialy increases startup time and memory consumption |

Reviewer 2: ```suggestion
| Configuration item | Description | Note | 
| ---------| ---- | ----|
| [`concurrently-init-stats`](/tidb-configuration-file.md#concurrently-init-stats-new-in-v810-and-v752), [`force-init-stats`](/tidb-configuration-file.md#force-init-stats-new-in-v657-and-v710), and [`lite-init-stats`](/tidb-configuration-file.md#lite-init-stats-new-in-v710) | Ensure the concurrent and comprehensive loading of table statistics during TiDB startup, which improves initial query optimization performance. | Might increase startup duration and memory usage. |
```
